### ISSUE TITLE
new unit in ASSIGN never retries if network was down

### ISSUE THREAD

**Issue by kbernhagen** [FoldingAtHome/fah-client-bastet#50]
macOS 12.6.1
fah 8.1.3 dev 2022-10-28

A unit that wants to retry apparently never does.

To reproduce:

* set client to finish and let all work upload
* disable networking
* unpause
* a new WU is created in ASSIGN state
* because network is down, the unit sets retry 2 seconds
* re-enable network
* still no retry
* pause and unpause
* still no retry

Workaround:

* restart client

Possible fix:

When a unit wants to retry in n seconds, create a timer for retry.

```
  app.getEventBase().setTimeout([this] () {
      app.getUnits().triggerUpdate(true);
    }, n);
```

A timed unit triggerNext() would be invalid if the user dumps a unit that is waiting. A triggerUpdate() is harmless if the unit is deleted.

Status JSON for stuck unit:

```
{
  "units": [
    {
      "number": 2,
      "cpus": 4,
      "gpus": [],
      "state": "ASSIGN",
      "progress": "NaN",
      "paused": false,
      "id": "S7qY3YGsLDd-49IupwtU3j0JyjYZuw1CyyvS2LzKYIU",
      "delay": 2,
      "wait": "2022-11-01T09:41:37Z",
      "retries": 1
    }
  ],
  "config": {
    "user": "Anonymous",
    "team": 0,
    "passkey": "",
    "checkpoint": 15,
    "on_idle": false,
    "paused": false,
    "finish": false,
    "cause": "any",
    "cpus": 4,
    "gpus": {
      "gpu:09:00:00": {"enabled": false}
    },
    "peers": [],
    "key": 0
  },
  "info": {
    "version": "8.1.3",
    "os": "macosx",
    "cpu": "amd64",
    "cpus": 32,
    "gpus": {
      "gpu:09:00:00": {"type": "amd", "description": "Vega 20"}
    },
    "id": "hfkN/qHYAn4UY+XKlxolARCiDfARnSBKdCILxjUW0Vs="
  }
}
```

Log:

[log.txt](https://github.com/FoldingAtHome/fah-client-bastet/files/9908808/log.txt)

**Comment by jcoffland** [comment]
I believe I found the problem.  The client failed to resolve the domain name for the AS and threw an exception in from within the call to ``OutgoingRequest::send()`` from ``Unit::assign()``.  The exception is caught in ``Unit::next()`` and ``Unit::response()`` never gets called.  This leaves ``Unit::pr`` set.  So two seconds later when ``Unit::next()`` is called it calls to ``Unit::assign()`` but nothing happens because ``Unit::pr`` is still set.

I've fixed this by always clearing ``Unit::pr`` in ``Unit::retry()``.
